ports/esp32/machine_i2s.c

```diff
@@ -0,0 +1,440 @@
+    }
+    i2s_config.fixed_mclk = self->apllrate;
+
+    // uninstall I2S driver when changes are being made to an active I2S peripheral
+    if (self->used) {
+        i2s_driver_uninstall(self->id);
+    }
+
+    esp_err_t ret = i2s_driver_install(self->id, &i2s_config, 0, NULL);
+    switch (ret) {
+        case ESP_OK:
+            break;
+        case ESP_ERR_INVALID_ARG:
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("I2S driver install:  Parameter error"));
+            break;
+        case ESP_ERR_NO_MEM:
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("I2S driver install:  Out of memory"));
+            break;
+        default:
+            // this error not documented in ESP-IDF
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("I2S driver install:  Undocumented error")); 
+            break;
+    }
+
+    i2s_pin_config_t pin_config;
+    pin_config.bck_io_num = self->bck;
+    pin_config.ws_io_num = self->ws;
+    pin_config.data_out_num = self->sdout;
+    pin_config.data_in_num = self->sdin;
+
+    ret = i2s_set_pin(self->id, &pin_config);
+    switch (ret) {
+        case ESP_OK:
+            break;
+        case ESP_ERR_INVALID_ARG:
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("I2S set pin:  Parameter error"));
+            break;
+        case ESP_FAIL:
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("I2S set pin:  IO error"));
+            break;
+        default:
+            // this error not documented in ESP-IDF
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("I2S set pin:  Undocumented error")); 
+            break;
+    }
+
+    self->used = true;
+}
+
+/******************************************************************************/
+// MicroPython bindings for I2S
+STATIC void machine_i2s_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    machine_i2s_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_printf(print, "I2S(id=%u, bck=%d, ws=%d, sdout=%d, sdin=%d\n"
+            "standard=%u, mode=%u,\n"
+            "dataformat=%u, channelformat=%u,\n"
+            "samplerate=%d,\n"
+            "dmacount=%d, dmalen=%d,\n"
+            "apllrate=%d)",
+            self->id, self->bck, self->ws, self->sdout, self->sdin,
+            self->standard, self->mode,
+            self->dataformat, self->channelformat,
+            self->samplerate,
+            self->dmacount, self->dmalen,
+            self->apllrate
+            );
+}
+
+STATIC mp_obj_t machine_i2s_make_new(const mp_obj_type_t *type, size_t n_pos_args, size_t n_kw_args, const mp_obj_t *args) {
+    mp_arg_check_num(n_pos_args, n_kw_args, 1, MP_OBJ_FUN_ARGS_MAX, true);
+
+    machine_i2s_obj_t *self;
+
+    // note: it is safe to assume that the arg pointer below references a positional argument because the arg check above
+    //       guarantees that at least one positional argument has been provided
+    i2s_port_t i2s_id = mp_obj_get_int(args[0]);
```

Comment:
`i2s_port_t i2s_id` : is this a port or an id??

Reply:
"id" identifies an I2S peripheral.  The ESP32 has two.   The term "id" is used consistently in the MicroPython Machine module classes, to identify a hardware port or channel.  Take a look at docs and code for comparable classes like SPI and I2C and you will see how this is used.   If you go to the top of this code file there is a comment about naming conventions.  It is confusing, but consistent.  Consistent is important.

Reply:
What I took away from your description at the top is that ids and ports are different things. But you define variables called "id" using a type called "port". Isn't that confusing?
If id==port and they end up being two names for the same values then I would change the comment at the top and say that this driver uses "id", i.e., the MicroPython name, and that there is an identity mapping from "id" to ESP-IDF's "port". Thereafter use variables with "id" and types with "id".
You could also use "port", then you'd have `i2s_port_t i2s_port = mp_obj_get_int(args[0]);`

Reply:
Thanks explaining where the comment is confusing.  I'll update the comment as per your suggestions.